File: genql/compilers/schema_compiler.py
"""
genql.compilers.schema_compiler
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Compiles SCHEMA intents into JSON Schema objects (as Python dicts).

Supported body keys
-------------------
title : str
    Human-readable schema title.
description : str  (optional)
    Description of the schema.
properties : dict[str, dict]
    Mapping of property name → property definition.
    Each definition may include ``type``, ``description``,
    ``enum``, ``default``, and any other JSON Schema keywords.
required : list[str]  (optional)
    Names of required properties.
additional_properties : bool  (optional, default False)
    Whether unknown properties are allowed.
"""

from __future__ import annotations

import json
from typing import Any, Dict, List

from genql.compilers.base import Compiler
from genql.intent import Intent, IntentKind


class SchemaCompiler(Compiler):
    """Translates :attr:`IntentKind.SCHEMA` intents to JSON Schema."""

    @property
    def target_language(self) -> str:
        return "json_schema"

    @property
    def supported_kinds(self) -> List[IntentKind]:
        return [IntentKind.SCHEMA]

    def compile(self, intent: Intent) -> str:
        body = intent.body
        schema: Dict[str, Any] = {
            "$schema": "https://json-schema.org/draft/2020-12/schema",
            "title": body.get("title", intent.name),
            "type": "object",
        }
        if "description" in body:
            schema["description"] = body["description"]

        properties = body.get("properties") or {}
        if properties:
            schema["properties"] = properties

        required = body.get("required") or []
        if required:
            schema["required"] = required

        schema["additionalProperties"] = body.get("additional_properties", False)

        return json.dumps(schema, indent=2)

File: genql_ruby/lib/genql/compilers/schema_compiler.rb
# genql/compilers/schema_compiler
#
# Compiles SCHEMA intents into JSON Schema objects (as JSON strings).
#
# Supported body keys
# -------------------
# title                 : String  — human-readable schema title
# description           : String  — description (optional)
# properties            : Hash    — property name → property definition
# required              : Array   — names of required properties (optional)
# additional_properties : Boolean — whether unknown properties are allowed (default false)

require "json"
require_relative "base"
require_relative "../intent"

module Genql
  module Compilers
    class SchemaCompiler < Compiler
      def target_language
        "json_schema"
      end

      def supported_kinds
        [IntentKind::SCHEMA]
      end

      def compile(intent)
        body = intent.body

        schema = {
          "$schema"    => "https://json-schema.org/draft/2020-12/schema",
          "title"      => body_get(body, "title") || intent.name,
          "type"       => "object"
        }

        description = body_get(body, "description")
        schema["description"] = description if description

        properties = body_get(body, "properties") || {}
        schema["properties"] = properties unless properties.empty?

        required = body_get(body, "required") || []
        schema["required"] = required unless required.empty?

        additional = body_get(body, "additional_properties")
        schema["additionalProperties"] = additional.nil? ? false : additional

        JSON.pretty_generate(schema)
      end

      private

      def body_get(body, key)
        body[key.to_sym] || body[key.to_s]
      end
    end
  end
end
